src/library_pipefs.js

```diff
@@ -0,0 +1,200 @@
+
+      var readableStream = FS.createStream({
+        path: rName,
+        node: rNode,
+        flags: FS.modeStringToFlags('r'),
+        seekable: false,
+        stream_ops: PIPEFS.stream_ops
+      });
+      rNode.stream = readableStream;
+
+      var writableStream = FS.createStream({
+        path: wName,
+        node: wNode,
+        flags: FS.modeStringToFlags('w'),
+        seekable: false,
+        stream_ops: PIPEFS.stream_ops
+      });
+      wNode.stream = writableStream;
+
+      return {
+        readable_fd: readableStream.fd,
+        writable_fd: writableStream.fd
+      };
+    },
+    stream_ops: {
+      poll: function (stream) {
+        var pipe = stream.node.pipe;
+
+        if ((stream.flags & {{{ cDefine('O_ACCMODE') }}}) === {{{ cDefine('O_WRONLY') }}}) {
+          return ({{{ cDefine('POLLWRNORM') }}} | {{{ cDefine('POLLOUT') }}});
+        } else {
+          if (pipe.buckets.length > 0) {
+            for (var i = 0; i < pipe.buckets.length; i++) {
+              var bucket = pipe.buckets[i];
+              if (bucket.offset - bucket.roffset > 0) {
+                return ({{{ cDefine('POLLRDNORM') }}} | {{{ cDefine('POLLIN') }}});
+              }
+            }
+          }
+        }
+
+        return 0;
+      },
+      ioctl: function (stream, request, varargs) {
+        return ERRNO_CODES.EINVAL;
+      },
+      read: function (stream, buffer, offset, length, position /* ignored */) {
+        var pipe = stream.node.pipe;
+        var currentLength = 0;
+
+        pipe.buckets.forEach(function (bucket) {
+          currentLength += bucket.offset - bucket.roffset;
+        });
+
+        assert(buffer instanceof ArrayBuffer || ArrayBuffer.isView(buffer));
+        var data = buffer.subarray(offset, offset + length);
+
+        if(length <= 0) {
+          return 0;
+        }
+        if (currentLength == 0) {
+          // Behave as if the read end is always non-blocking
+          throw new FS.ErrnoError(ERRNO_CODES.EAGAIN);
+        }
+        var toRead = Math.min(currentLength, length);
+
+        var totalRead = toRead;
+        var toRemove = 0;
+
+        for (var i = 0; i < pipe.buckets.length; i++) {
+          var currBucket = pipe.buckets[i];
+          var bucketSize = currBucket.offset - currBucket.roffset;
+
+          if (toRead <= bucketSize) {
+            var tmpSlice = currBucket.buffer.subarray(currBucket.roffset, currBucket.offset);
+            if (toRead < bucketSize) {
+              tmpSlice = tmpSlice.subarray(0, toRead);
+              currBucket.roffset += toRead;
+            } else {
+              toRemove++;
+            }
+            data.set(tmpSlice);
+            break;
+          } else {
+            var tmpSlice = currBucket.buffer.subarray(currBucket.roffset, currBucket.offset);
+            data.set(tmpSlice);
+            data = data.subarray(tmpSlice.byteLength);
+            toRead -= tmpSlice.byteLength;
+            toRemove++;
+          }
+        }
+
```

Comment:
I think we could add something like this to prevent deallocating all buffers and leave one bucket alive:

    if (toRemove && toRemove == pipe.buckets.length) {
        toRemove--;
        pipe.buckets[toRemove].offset = 0;
        pipe.buckets[toRemove].roffset = 0;
    }